### PR TITLE
[multistage][bugfix] improve sort copy rule

### PR DESCRIPTION
**Pull request by walterddr**
logical sort should not be copied when there's no fetch or fetch limit is too large.

this solves part 1 of https://github.com/apache/pinot/issues/12228

follow up separately later
1. when we support k-merge sort we should allow sort-copy again for distributed local sort.
2. make fetch limit configurable.

### CODE REVIEW

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotSortExchangeCopyRule.java [hunk 1]
```diff
@@ -42,6 +42,7 @@ public class PinotSortExchangeCopyRule extends RelRule<RelRule.Config> {
 
   public static final PinotSortExchangeCopyRule SORT_EXCHANGE_COPY =
       PinotSortExchangeCopyRule.Config.DEFAULT.toRule();
+  private static final int DEFAULT_SORT_EXCHANGE_COPY_THRESHOLD = 10_000;
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotSortExchangeCopyRule.java:45]: 10_000 is a reasonable heuristic until we get a config/hint.

Comment (xiangfu0) [reply pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotSortExchangeCopyRule.java:45]: is it worth to make this configurable as a query param?

**Review by xiangfu0 — APPROVED** [review]

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotSortExchangeCopyRule.java (hunk 1 above)
Comment (walterddr) [reply pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotSortExchangeCopyRule.java:45]: probably. but i haven't decided whether to use a hint or use a query param. so for now i will keep this as a default and if we need to support configuration we will do that then.

**Comment by gortiz** [comment]
> or fetch limit is too large.

Recently, we found cases where users were using very large limit, and this change backfired us. What is the advantage of removing the sort when the limit is larger than 10k?

**Pull request merged**